### PR TITLE
HPCC-14800 Add ID suggestions to unknown id errors

### PR DESCRIPTION
**Pull request by jamienoss**
default is on - deactivated via flag --auto-id-suggestions

Signed-off-by: James Noss james.noss@lexisnexis.com

### CODE REVIEW

**Comment by jamienoss** [comment]
@ghalliday

**Comment by jamienoss** [comment]
FYI https://github.com/hpcc-systems/HPCC-Platform/pull/8130

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -12003,3 +12362,186 @@ IHqlExpression *HqlGram::doParse()
     return NULL;
 }
 
+/*The following confusion  matrix holds the cost of transpositions and substitutions for weighted-Damerau-Levenshtein Distance.
```
Comment (jamienoss) [inline ecl/hql/hqlgram2.cpp]: This code actually belongs to https://github.com/hpcc-systems/HPCC-Platform/pull/8130 but wasn't sure on how to incorporate since the code in that PR is within the plugins?

File: ecl/eclcc/eclcc.cpp [hunk 2]
```diff
@@ -2063,6 +2068,10 @@ bool EclCC::parseCommandLineOptions(int argc, const char* argv[])
         {
             setDebugOption("syntaxCheck", tempBool);
         }
+        else if (iter.matchFlag(tempBool, "--auto-id-suggestions"))
```
Comment (ghalliday) [inline ecl/eclcc/eclcc.cpp]: trivial: Can directly use optAutoIdSuggestions in place of tempBool

File: ecl/hql/hqlexpr.cpp [hunk 3]
```diff
@@ -8147,6 +8208,114 @@ IHqlExpression *CHqlRemoteScope::clone(HqlExprArray &newkids)
...
+            isRepoSymOk = true;
+        else
+        {
+            IFileContents * contents = repoSym->queryDefinitionText();
+            if (contents && (contents->length() != 0))
+                isRepoSymOk = true;
+        }
+    }
+
+    OwnedHqlExpr nearest(HqlGram::findNearestSymbol(distance = -1, ret, isRetOk, retDistance, repoSym, isRepoSymOk, repoSymDistance));
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: Style: Assigning in a parameter to a function is really ugly and confusing.

File: ecl/hql/hqlgram2.cpp [hunk 4]
```diff
@@ -3333,6 +3352,79 @@ IHqlExpression *HqlGram::lookupSymbol(IHqlScope * scope, IIdAtom * searchName)
...
 }
 
+const char * HqlGram::composeSymbolSuggestionMsg(IHqlExpression * symbol, StringBuffer & msg)
+{
+    if (symbol)
+        msg.append(" - did you perhaps mean \"").append(symbol->queryName()->queryStr()).append("\"?");
+    return msg.str();
+}
+
+IHqlExpression * HqlGram::findNearestSymbol(unsigned & editDistance, const HqlExprArray & symbolArray, const unsigned * distanceArray, unsigned penalty)
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: This function doesn't belong in this class.  I don't think it uses any of the class members.

File: ecl/hql/hqlgram2.cpp [hunk 5]
```diff
@@ -3333,6 +3352,79 @@ IHqlExpression *HqlGram::lookupSymbol(IHqlScope * scope, IIdAtom * searchName)
...
+        {
+            minDistance = tempDistance;
+            editDistance = distanceArray[i];
+            nearest = &symbolArray.item(i);
+        }
+    }
+    return LINK(nearest);
+}
+
+IHqlExpression * HqlGram::findNearestSymbol(unsigned & minDistance, IHqlExpression * s1, bool ok1, unsigned distance1, IHqlExpression * s2, bool ok2, unsigned distance2)
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: See general comments.  With a class that kept the best so-far I suspect this is redundant.

File: ecl/hql/hqlgram2.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: This PR should be based on that one, and you should add a dependency on hql.so on stringlib.so.  The string lib plugins are part of the core platform/

**Comment by ghalliday** [comment]
First some general comments/impressions.

The first is that there is more code than I would expect.  In particular some of it is duplicated quite a lot. E.g., 
1) the snippet:

        unsigned distance = HqlGram::computeEditDistance(fieldName, cur->queryId());
        if (distance == 0)
            continue;
        else if (distance < editDistance && distance <= MAX_SYMBOL_DISTANCE)
        {
            ret = cur;
            editDistance = distance;
        }

E.g., in this instance I would introduce a class that held a (IIdAtom and a distance), and had a member to check if a new id should replace it.  I.e.

    bestMatch.checkCandidate(cur->queryId(), penalty);

2) The code in CRemoteScope is much too involved.  It should be doing something like calling ensureSymbolsDefined, and then working through the symbol table.  There is no need to actually resolve the definitions, and in fact that may introduce a large amount of processing if it requires the definitions to be processed.

3) Personally I wouldn't include the code inside CHqlRemoteScope and CHqlRecord.  It should be posssible to implement it as an external function for an IHqlScope by calling getSymbols() (after ensureSymbolsDefined()), and iterating through the symbols.   I would extend CHqlRecord and IHqlSimpleScope to provide a similar function.

**Comment by ghalliday** [comment]
@jamienoss See some initial comments on the structure and a few comments on the code.

**Comment by ghalliday** [comment]
The PR is also missing the test cases.

File: ecl/eclcc/eclcc.cpp (hunk 2 above)
Comment (jamienoss) [inline ecl/eclcc/eclcc.cpp]: Yep, thanks,  I forgot to change this, it was left over from when I had the internal bool option named optNoAutoIdSugestion and thus had to invert.

**Comment by richardkchapman** [comment]
@ghalliday Ready for re-review?

File: ecl/hql/hqlexpr.cpp [hunk 6]
```diff
@@ -6614,6 +6615,27 @@ IHqlExpression *CHqlRecord::lookupSymbol(IIdAtom * fieldName)
     return ret;
 }
 
+IHqlExpression * CHqlRecord::lookupNearestSymbol(IIdAtom * fieldName)
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I don't think the versions without the NearestSymbol parameter are called from anywhere, so can be deleted.

File: ecl/hql/hqlexpr.cpp [hunk 7]
```diff
@@ -6614,6 +6615,27 @@ IHqlExpression *CHqlRecord::lookupSymbol(IIdAtom * fieldName)
...
+{
+    NearestSymbol nearest(fieldName);
+    return LINK(lookupNearestSymbol(fieldName, nearest));
+}
+
+IHqlExpression * CHqlRecord::lookupNearestSymbol(IIdAtom * fieldName, NearestSymbol & nearest)
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: Better to not return the symbol - since it is always available from the parameter.  One minor danger is that lookupSymbol() returns a linked expression, while this doesn't - so potential scope for confusion.

File: ecl/hql/hqlexpr.cpp [hunk 8]
```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
...
+}
+
+void NearestSymbol::compare(IHqlExpression * comparison)
+{
+    if (!comparison || nearest == comparison)
+        return;
+
+    unsigned tempDistance = computeEditDistance(comparison->queryId());
+
+    if (tempDistance == 0 || tempDistance >= MAX_SYMBOL_DISTANCE)
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: Why the check for tempDistance == 0 - that means that it can never pick the best value for weightedDamerauLevenshtein.

File: ecl/hql/hqlexpr.cpp [hunk 9]
```diff
@@ -8203,7 +8345,19 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
     {
         StringBuffer msg("Definition for ");
         msg.append(str(searchName)).append(" contains no text");
-        throw createError(ERR_EXPORT_OR_SHARE, msg.str(), filename, 0, 0, 1);
+        if(ctx.queryParseContext().autoIdSuggestions)
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I don't think this is likely to be a useful improvement.  Do you have examples where it is?

File: ecl/hql/hqlexpr.cpp [hunk 10]
```diff
@@ -8156,7 +8288,17 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
     if (resolvedSym && resolvedSym->getOperator() == no_processing)
     {
         if (lookupFlags & LSFrequired)
-            throwRecursiveError(searchName);
+        {
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I don't think this is likely to improve the error message - do you have examples where it does?

File: ecl/hql/hqlexpr.cpp [hunk 11]
```diff
@@ -8147,6 +8194,91 @@ IHqlExpression *CHqlRemoteScope::clone(HqlExprArray &newkids)
     return this;
 }
 
+IHqlExpression * CHqlRemoteScope::lookupNearestSymbol(IIdAtom * searchName, unsigned lookupFlags, HqlLookupContext & ctx, NearestSymbol & nearest)
+{
+    preloadSymbols(ctx, false);
+
+    unsigned penalty = 0;
+    NearestSymbol tempNearest(searchName);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I don't think there is any need to check the resolved member - all the information is in the base implementation.

File: ecl/hql/hqlexpr.cpp [hunk 12]
```diff
@@ -8147,6 +8194,91 @@ IHqlExpression *CHqlRemoteScope::clone(HqlExprArray &newkids)
...
+    {
+        nearest.compare(tempNearest, penalty);
+        if (nearest.bestMatchFound())
+            return nearest.querySymbol();
+    }
+
+    return nearest.querySymbol();
+    //MORE: the logic below rquires parseAttribute to open to edit distance searches
+
+/*
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: unused code should be deleted.

File: ecl/hql/hqlexpr.cpp [hunk 13]
```diff
@@ -8229,6 +8383,13 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
             resolved->removeSymbol(searchName);
         StringBuffer msg("Definition must contain EXPORT or SHARED value for ");
         msg.append(str(searchName));
+        if(ctx.queryParseContext().autoIdSuggestions)
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: not sure this is useful...

File: ecl/hql/hqlexpr.cpp [hunk 14]
```diff
@@ -8238,9 +8399,9 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
     assertex(symbol);
     symbol->setRepositoryFlags(repositoryFlags);
 
-    if (repositoryFlags&ob_sandbox)
+    if (repositoryFlags & ob_sandbox)
     {
-        if (ctx.errs)
+        if (ctx.errs)//MORE: should this error report searchName and thus a suggestion also?
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: no.

File: ecl/hql/hqlexpr.cpp [hunk 15]
```diff
@@ -8471,6 +8632,36 @@ inline bool canMergeDefinition(IHqlExpression * expr)
     return true;
 }
 
+IHqlExpression * CHqlMergedScope::lookupNearestSymbol(IIdAtom * searchId, unsigned lookupFlags, HqlLookupContext & ctx, NearestSymbol & nearest)
+{
+    CriticalBlock block(cs);
+    NearestSymbol tempNearest(searchId);
+    unsigned penalty = 0;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: This logic isn't correct.  I think it should be looping through mergedScopes, calling lookupNearest on each.

File: ecl/hql/hqlexpr.cpp [hunk 16]
```diff
@@ -9143,6 +9338,122 @@ IHqlExpression * CHqlVirtualScope::lookupBaseSymbol(IHqlExpression * & definitio
...
+    IHqlExpression * scopeSymbol = CHqlScope::lookupNearestSymbol(searchName, lookupFlags, ctx, scopeNearest);
+
+    IHqlExpression * definitionModule = this;
+    NearestSymbol baseNearest(searchName);
+    IHqlExpression * baseSymbol = lookupNearestBaseSymbol(definitionModule, searchName, lookupFlags, ctx, baseNearest);
+
+    if (!scopeSymbol && !baseSymbol)
+        return nearest.querySymbol();//NULL
+
+    bool isScopeSymbolOK = false;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: This code doesn't make much sense.  Please ask if you don't know what a class represents.

File: ecl/hql/hqlexpr.cpp [hunk 17]
```diff
@@ -9372,6 +9684,63 @@ void CHqlForwardScope::defineSymbol(IHqlExpression * expr)
         CHqlVirtualScope::defineSymbol(expr);
 }
 
+IHqlExpression * CHqlForwardScope::lookupNearestSymbol(IIdAtom * searchName, unsigned lookupFlags, HqlLookupContext & ctx, NearestSymbol & nearest)
+{
+    NearestSymbol tempNearest(searchName);
+    unsigned penalty = 0;
+    IHqlExpression * resolvedSym = resolved->lookupNearestSymbol(searchName, lookupFlags, ctx, tempNearest);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: Again, I don't think there is any need to look at resolved.

File: ecl/hql/hqlexpr.cpp [hunk 18]
```diff
@@ -9477,6 +9846,25 @@ CHqlMultiParentScope::CHqlMultiParentScope(IIdAtom * _name, ...)
     va_end(args);
 }
 
+IHqlExpression * CHqlMultiParentScope::lookupNearestSymbol(IIdAtom * searchName, unsigned lookupFlags, HqlLookupContext & ctx, NearestSymbol & nearest)
+{
+    unsigned penalty = 0;
+    NearestSymbol tempNearest(searchName);
+    CHqlScope::lookupNearestSymbol(searchName, lookupFlags, ctx, tempNearest);
+    nearest.compare(tempNearest, penalty);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I'm unclear about this version of compare which takes penalty as a reference.  Please can you clarify the idea behind it.

File: ecl/hql/hqlexpr.cpp [hunk 19]
```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
     return transformer.transform(expr);
 }
 
+/*The following confusion  matrix holds the cost of transpositions and substitutions for weighted-Damerau-Levenshtein Distance.
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: This should be deleted, and this PR should be based on the branch that contains the code adding it to stringlib.

File: ecl/hql/hqlexpr.cpp [hunk 20]
```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
...
+    case weightedDamerauLevenshtein :
+        return damerauLevenshteinDistance(strlen(searchNameLowerStr), searchNameLowerStr, strlen(comparisonLowerStr), comparisonLowerStr, defaultConfusionMatrix, defaultConfusionMatrixLength) + 1;
+    };
+    return -1;
+}
+
+IHqlExpression * NearestSymbol::branchSearch(IHqlScope * scope, unsigned lookupFlags, HqlLookupContext & ctx, unsigned & penalty)
+{
+    NearestSymbol tempNearest(searchName);
+    scope->lookupNearestSymbol(searchName, lookupFlags, ctx, tempNearest);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: I doubt this will compile.

File: ecl/hql/hqlexpr.cpp [hunk 21]
```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
...
+
+    if (nearest == comparison.nearest)
+        return;
+
+    unsigned unsignedDistance = comparison.distance == (unsigned)(-1) ? comparison.distance : comparison.distance + penalty;
+    if (unsignedDistance < distance)
+    {
+        distance = comparison.distance;
+        nearest.set(comparison.nearest);
+        penalty++;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: why increase penalty?  See comment elsewhere.

File: ecl/hql/hqlexpr.cpp [hunk 22]
```diff
@@ -16188,6 +16618,292 @@ IHqlExpression * annotateIndexBlobs(IHqlExpression * expr)
...
+        return false;
+
+    switch(metric)
+    {
+    case optimalStringAlignment :
+    case damerauLevenshtein :
+    {
+        if (distance == 1)
+            return true;
+    }
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: missing break.  

style: simplest code would be

  case optimalStringAlignment :
  case damerauLevenshtein :
      return distance == 1;

File: ecl/hql/hqlexpr.hpp [hunk 23]
```diff
@@ -1201,6 +1209,37 @@ interface IHqlNamedAnnotation : public IHqlAnnotation
...
+    void compare(IHqlExpression * expr);
+    void compare(NearestSymbol & comparison);
+    void compare(NearestSymbol & comparison, unsigned & penalty);
+    bool bestMatchFound() const;
+    IHqlExpression * branchSearch(IHqlSimpleScope * scope, unsigned & penalty);
+    IHqlExpression * branchSearch(IHqlScope * scope, unsigned lookupFlags, HqlLookupContext & ctx, unsigned & penalty);
+    IHqlExpression * querySymbol() const { return nearest; }
+
+private :
+    Owned<IIdAtom> searchName;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.hpp]: Should just be an IIdAtom, atoms are not actively link counted.

File: ecl/hql/hqlexpr.hpp [hunk 24]
```diff
@@ -1201,6 +1209,37 @@ interface IHqlNamedAnnotation : public IHqlAnnotation
...
+{
+public :
+    enum metricType { optimalStringAlignment, damerauLevenshtein, weightedDamerauLevenshtein };
+
+    NearestSymbol(IIdAtom * _searchName, metricType _metric) : distance(-1), metric(_metric)
+    {
+        searchName.set(_searchName);
+    }
+    NearestSymbol(IIdAtom * _searchName) : NearestSymbol(_searchName, weightedDamerauLevenshtein) {}
+    ~NearestSymbol() { searchName.clear(); nearest.clear(); }
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.hpp]: minor: no need for the destructor.

File: ecl/hql/hqlexpr.ipp [hunk 25]
```diff
@@ -39,6 +39,7 @@
 #include "javahash.hpp"
 #include "defvalue.hpp"
 #include "hqlexpr.hpp"
+#include "hqlgram.hpp"
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.ipp]: Why is this needed?  If it is the definition of the function I would prefer the function to be defined in hqlexpr.cpp.  The visibility/dependence on the parser should be as small as possible.

File: ecl/hql/hqlgram2.cpp [hunk 26]
```diff
@@ -3346,6 +3358,147 @@ unsigned HqlGram::getExtraLookupFlags(IHqlScope * scope)
...
+
+    unsigned penalty = 0;
+    try
+    {
+        // If there is a temporary scope, we only look up in that (and it must exist!).
+        if (dotScope)
+        {
+            if (dotScope->getOperator() == no_enum)
+            {
+                NearestSymbol tempNearest(searchName);
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: The use of all these temporary NearestSymbol objects leads me to suspect that the logic isn't quite right.  Should the penalty be being passed in to the lookupNearestSymbol function instead?  When should the penalty be incremented?  

How should the penalty decide on the best match?  You add it to the distance - but should it instead work more as a tie breaker?  If the penalty does affect the distance, then it should be passed in to the lookup function because it means the best match can be short-circuited more quickly.

File: ecl/hql/hqlgram2.cpp [hunk 27]
```diff
@@ -11006,14 +11201,24 @@ void HqlGram::syntaxError(const char *s, int token, int *expected)
     {
         msg.append("Unknown identifier");
         if (yytext && *yytext)
+        {
             msg.append(" \"").append(yytext).append('\"');
+            if(lookupCtx.queryParseContext().autoIdSuggestions)
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: This logic probably needs improving.  You should only report a suggestion of an alternative identifier if a valid predefined identifier is valid at this point.

If no identifier is valid (and if it is) you should be auto-suggesting against the list of keywords that are valid next.  Leave that to a secondary PR though - it would be better to get this one merged first.

File: ecl/hql/hqllex.l [hunk 28]
```diff
@@ -64,6 +64,94 @@ extern void hex2str(char * target, const char * digits, unsigned len);
...
+        return '_';
+    else if (n > 10  && n < 37)
+        return 'a' + n - 11;
+    else
+        throwUnexpected();
+}
+
+static unsigned scrambledIdCount = 0;
+
+class DummyAtom : public CInterfaceOf<IIdAtom>
```
Comment (ghalliday) [inline ecl/hql/hqllex.l]: This is almost certainly not wanted.  You should create an IdAtom in the same way as normal.

File: ecl/hql/hqllex.l [hunk 29]
```diff
@@ -64,6 +64,94 @@ extern void hex2str(char * target, const char * digits, unsigned len);
...
+    virtual void                setHash(unsigned _hash) { hash = _hash; }
+    virtual IAtom *             queryLower() const { return lower; }
+
+private:
+    unsigned hash;
+    char * text;
+    IAtom * lower;
+};
+
+static IIdAtom * createScrambledIdAtom(const IIdAtom * idToScramble, unsigned nCharToScramble, unsigned probability, unsigned maxIdsToScramble)
```
Comment (ghalliday) [inline ecl/hql/hqllex.l]: This should be in a different file - e.g., hqlparse.cpp.  It is confusing the common case code with some debug test code.

File: ecl/hql/hqllex.l [hunk 30]
```diff
@@ -64,6 +64,94 @@ extern void hex2str(char * target, const char * digits, unsigned len);
...
+    }
+
+    if (nCharToScramble > 256)
+        nCharToScramble = 256;
+
+    unsigned length = strlen(str(idToScramble));
+    if (nCharToScramble > length)
+        nCharToScramble = length;
+
+    char ret[length+1];
```
Comment (ghalliday) [inline ecl/hql/hqllex.l]: It is probably ok, but I'm not keen on this.  (Allocating something of unknown length on the stack.)

File: ecl/hql/hqllex.l [hunk 31]
```diff
@@ -228,6 +316,53 @@ int HqlLex::lookupIdentifierToken(YYSTYPE & returnToken, HqlLex * lexer, bool lo
         token = UNKNOWN_ID;
     }
 
+#ifdef TEST_ID_SUGGESTIONS
```
Comment (ghalliday) [inline ecl/hql/hqllex.l]: again this should be a function defined elsewhere.

**Comment by ghalliday** [comment]
@jamienoss back to you

File: ecl/hql/hqlexpr.cpp (hunk 20 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.cpp]: This has to be an odd github bug, as this is not the actual code, which is complete and does compile.

File: ecl/hql/hqlexpr.hpp [hunk 32]
```diff
@@ -1201,6 +1209,37 @@ interface IHqlNamedAnnotation : public IHqlAnnotation
...
+    void compare(NearestSymbol & comparison);
+    void compare(NearestSymbol & comparison, unsigned & penalty);
+    bool bestMatchFound() const;
+    IHqlExpression * branchSearch(IHqlSimpleScope * scope, unsigned & penalty);
+    IHqlExpression * branchSearch(IHqlScope * scope, unsigned lookupFlags, HqlLookupContext & ctx, unsigned & penalty);
+    IHqlExpression * querySymbol() const { return nearest; }
+
+private :
+    Owned<IIdAtom> searchName;
+    Owned<IHqlExpression> nearest;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.hpp]: You should probably be saving the id of the nearest.  Long term (e.g., when comparing against reserved words) it will need to be.

File: ecl/hql/hqlexpr.cpp (hunk 8 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.cpp]: The weighted distance has a +1 so that it can never be zero itself. Since a mixture of distances is not used this is ok.

File: ecl/hql/hqlexpr.cpp (hunk 6 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.cpp]: lookupNearestSymbol(IIdAtom * fieldName) is part of IHqlSimpleScope which CHqlRecord implements.

File: ecl/hql/hqlexpr.cpp (hunk 21 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.cpp]: So the general point here is that the searches can be branched and compared but with a preference to those that were given preference in lookupSymbol, i.e. scope etc, we discussed such things once before. So each branch would have its own set of penalties for it's set of searches, when returning its best match this will then undergo such preferential comparisons. Think of the whole search as a tree.

File: ecl/hql/hqlexpr.cpp (hunk 18 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.cpp]: ok, so you're right there's probably not much need for it in this instance. I think I had more to this routine that did utilize it but has been cut out. I could unwrap this instance, but the method itself is used in enough place to warrant not removing it and thus it being neater to still use it here. See the comment below for the general idea behind it.

File: ecl/hql/hqlexpr.ipp (hunk 25 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.ipp]: Good spot, thank you. You're right this isn't needed.

File: ecl/hql/hqlexpr.cpp (hunk 8 above)
Comment (jamienoss) [inline ecl/hql/hqlexpr.cpp]: The weighted versions were offset by +1. Since comparisons between diff metrics is not conducted this is ok.

File: ecl/hql/hqlgram2.cpp (hunk 27 above)
Comment (jamienoss) [inline ecl/hql/hqlgram2.cpp]: Cool, good point, thanks.

**Comment by jamienoss** [comment]
@ghalliday I've made changes. Your comments that I haven't addressed I will discuss with you on flowdock.

File: ecl/hql/hqlexpr.cpp (hunk 6 above)
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: Yes... which nothing calls.

**Comment by ghalliday** [comment]
@jamienoss.  It looks like you have made very few changes when the changes are complete.
